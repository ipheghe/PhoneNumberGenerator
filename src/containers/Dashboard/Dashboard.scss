@import '../../common/functions.scss';

.container {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
  height: 100%;
  padding: rem(20px) rem(100px);
  margin: rem(100px) 0;

  .number__input__section {
    position: relative;
    display: flex;
    justify-content: flex-start;
    width: 100%;
    padding: 0;

    .number__details {
      font-weight: bold;
  
      span {
        font-weight: normal;
      }
    }

    .download__btn {
      position: absolute;
      background-color: #57CCCb;
      cursor: pointer;
      margin: auto 0;
      padding: 10px;
      right: 0;
      top: 40%;

      a {
        list-style-type: none;
        text-decoration: none;
        color: #FFF;
      }
    }
  }

  .dropdown__dashboard {
    align-self: flex-end;
    margin-bottom: rem(10px);
    margin-top: rem(30px);
  }

  .empty_table__text {
    display: flex;
    flex-direction: column;
    margin-top: rem(50px);
    align-items: center;
  }
}

@media screen and (max-width: rem(450px)) {
  .container {
    margin: rem(30px) 0;
    padding: rem(20px) rem(10px);

    .number__input__section {
      margin-top: rem(30px);
      flex-direction: column;

      .download__btn {
        position: static;
        max-width: rem(100px);

      }
    }

    .empty_table__text {
      margin: auto;
      margin-bottom: rem(20px);
      font-size: rem(18px);
      padding: rem(20px);
    }
  }
}
