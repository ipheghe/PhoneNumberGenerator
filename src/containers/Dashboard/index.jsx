// react imports
import React, { Component } from 'react';

// third-party libraries import
import _ from 'lodash';
import { CSVLink } from 'react-csv';

// styles
import './Dashboard.scss';

// components
import NumberInput from '../../components/NumberInput';
import Table from '../../components/Table';
import DropDown from '../../components/DropDown';
import { Consumer } from '../../components/Toast/ToastContext';

const configuration = {
  serial: {
    alias: '#',
    width: 'table-col-20',
  },
  id: { alias: 'id', width: 'table-col-30' },
  generatedNumbers: { alias: 'Generated Numbers' },
};

/**
 * Container component housing the Single-Page application
 *
 * @class Dashboard
 */
export class Dashboard extends Component {
  /**
   * Creates an instance of App
   *
   * @param {object} props - props
   * @memberof dashboard
   * @returns {void}
   */
  constructor(props) {
    super(props);

    this.state = {
      chosenNumber: 0,
      buttonText: 'Generate',
      errorMessage: '',
      data: [],
    };
  }

  /**
   * This method generates random phone numbers
   *
   * @return {void}  null
   */
  generatePhoneNumbers = () => {
    const { chosenNumber, data } = this.state;
    const { showMessage } = this.props;
    const formattedNumber = parseInt(chosenNumber, 10);
    let generatedNumber;
    let generatedData = [];

    for (let i = 1; i < formattedNumber + 1; i += 1) {
      generatedNumber = String(
        Math.floor(Math.random() * 900000000) + 100000000
      );
      const addZeroToNumbers = generatedNumber.padStart(10, '0');

      if (_.findIndex(data, { generatedNumbers: addZeroToNumbers }) === -1) {
        if (
          _.findIndex(generatedData, { generatedNumbers: addZeroToNumbers }) ===
          -1
        ) {
          generatedData.push({ generatedNumbers: addZeroToNumbers });
        }
      }
    }
    const lastId = data.sort((a, b) => b.id - a.id);
    const updatedGeneratedData = generatedData.map((item, index) => {
      item['id'] = lastId[0] ? lastId[0].id + index + 1 : 1;
      return item;
    });
    showMessage('Number(s) generated successfully', 'success');
    return updatedGeneratedData;
  };

  /**
   * This method validates input field after submitting
   *
   * @return {void}  null
   */
  validateField = () => {
    const { chosenNumber } = this.state;
    const { showMessage } = this.props;
    const formattedNumber = parseInt(chosenNumber, 10);
    if (chosenNumber % 1 !== 0) {
      showMessage('Number must be an integer', 'error');
      return false;
    }
    if (!formattedNumber || formattedNumber === 0) {
      showMessage('Please enter a number value', 'error');
      return false;
    }
    if (formattedNumber > 500) {
      showMessage('Number must be between 1 and 500', 'error');
      return false;
    }

    return true;
  };

  /**
   * This method handles number input change event
   *
   * @param {event} event - Sythentic event
   * @return {void}  null
   */
  handleChange = event => {
    event.preventDefault();
    this.setState({
      [event.target.name]: event.target.value,
    });
  };

  /**
   * This method handles text button Click to generate numbers
   *
   * @param {event} event - Sythentic event
   * @return {void}  null
   */
  handleButtonClick = event => {
    event.preventDefault();
    this.setState(
      {
        buttonText: 'Generating Numbers.....',
      },
      () => this.updateGeneratedNumbers()
    );
  };

  /**
   * This method updates generated numbers
   *
   * @return {void}  null
   */
  updateGeneratedNumbers = () => {
    const { data } = this.state;
    const isNumberFieldValidated = this.validateField();
    if (isNumberFieldValidated) {
      const generatedNumbersData = this.generatePhoneNumbers();
      const updatedData = [...data, ...generatedNumbersData];
      this.setState({
        data: updatedData,
        buttonText: 'Generate',
      });
    } else {
      this.setState({ buttonText: 'Generate' });
    }
  };

  /**
   * This method sorts generated numbers by selected option
   *
   * @param {event} event - Sythentic event
   * @param {string} sortOption - sort option
   * @return {void}  null
   */
  handleSortClick = (event, sortOption) => {
    event.preventDefault();

    let { data } = this.state;
    if (sortOption === 'highest') {
      data = _.orderBy(data, ['generatedNumbers'], ['desc']);
    }
    if (sortOption === 'lowest') {
      data = _.orderBy(data, ['generatedNumbers'], ['asc']);
    }
    if (sortOption === 'unique') {
      data = _.sortBy(data, ['id']);
    }

    this.setState(state => (state.data !== data ? { data } : null));
  };

  /**
   * This method renders table componenet or a text if no generated number
   *
   * @return {JSX}  JSX
   */
  renderTable = () => {
    const { data } = this.state;

    if (data && data.length > 0) {
      return (
        <>
          <DropDown
            className="dropdown__dashboard"
            handleSortClick={this.handleSortClick}
          />
          <Table
            data={data}
            config={configuration}
            headers={Object.keys(configuration)}
          />
        </>
      );
    }

    return (
      <div className="empty_table__text">
        <h1>No Generated Number Yet.</h1>
        <h1>Please enter a number to be generated</h1>
      </div>
    );
  };

  /**
   * This method renders details of max, min and total numbers
   *
   * @return {JSX}  JSX
   */
  renderNumberDetails = () => {
    const { data } = this.state;
    const maxNumber = data.length
      ? _.orderBy(data, ['generatedNumbers'], ['desc'])[0].generatedNumbers
      : 0;
    const minNumber = data.length
      ? _.orderBy(data, ['generatedNumbers'], ['asc'])[0].generatedNumbers
      : 0;

    return (
      <div className="number__details">
        <p>
          Total Numbers: <span>{data.length ? data.length : 0}</span>
        </p>
        <p>
          Max Number: <span>{maxNumber}</span>
        </p>
        <p>
          Minimum Numbers: <span>{minNumber}</span>
        </p>
      </div>
    );
  };

  /**
   * This method renders download button
   *
   * @return {JSX}  JSX
   */
  renderDownloadButton = () => {
    const { data } = this.state;
    if (data && data.length) {
      return (
        <button className="download__btn">
          <CSVLink data={data}>Download Data</CSVLink>
        </button>
      );
    }
  };

  /**
   * Renders Dashboard component
   *
   * @memberof Dashboard
   * @returns {JSX} jsx
   */
  render() {
    const { chosenNumber, buttonText, data } = this.state;

    return (
      <div className="container">
        <div className="number__input__section">
          {data && this.renderNumberDetails()}
          {this.renderDownloadButton()}
        </div>
        <NumberInput
          value={chosenNumber}
          onChange={this.handleChange}
          handleButtonClick={this.handleButtonClick}
          buttonText={buttonText}
        />
        {this.renderTable()}
      </div>
    );
  }
}

/**
 * This renders the dashboard component with context
 *
 * @param {object} props - props
 * @returns {JSX} JSX
 */
export default function DashboardWithContext(props) {
  return (
    <Consumer>
      {context => <Dashboard {...props} showMessage={context.showMessage} />}
    </Consumer>
  );
}
