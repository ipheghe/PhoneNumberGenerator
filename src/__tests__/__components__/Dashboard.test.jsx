// react libraries
import React from 'react';

// third-party libraries
import { shallow, mount } from 'enzyme';
import render from 'react-test-renderer';
import { spy } from 'sinon';
import _ from 'lodash';

// components
import { Dashboard } from '../../containers/Dashboard';

const event = {
  target: {
    name: 'chosenNumber',
    value: 3,
  },
  preventDefault: spy(),
};

const props = {
  showMessage: spy(),
};

const data = [
  { id: 1, generatedNumbers: 12345 },
  { id: 2, generatedNumbers: 11345 },
  { id: 3, generatedNumbers: 11345 },
];

describe('App component', () => {
  it('should match component snapshot', () => {
    const tree = render.create(<Dashboard {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('renders without crashing', () => {
    const mountWrapper = mount(<Dashboard {...props} />);
    expect(mountWrapper).toBeDefined();
    expect(mountWrapper.find('NumberInput').length).toBe(1);
    expect(mountWrapper.exists()).toBe(true);
  });

  it('should call handleChange method', () => {
    const mountWrapper = mount(<Dashboard {...props} />);
    const handleChangeSpy = spy(mountWrapper.instance(), 'handleChange');

    expect(mountWrapper.state().chosenNumber).toEqual(0);

    mountWrapper.instance().handleChange(event);
    expect(handleChangeSpy.called).toEqual(true);
    expect(mountWrapper.state().chosenNumber).toEqual(3);
  });

  it('should call handleButtonClick method', () => {
    const mountWrapper = shallow(<Dashboard {...props} />);
    const handleButtonClickSpy = spy(mountWrapper.instance(), 'handleButtonClick');

    expect(mountWrapper.state().buttonText).toEqual('Generate');

    mountWrapper.instance().handleButtonClick(event);
    expect(handleButtonClickSpy.called).toEqual(true);
    expect(mountWrapper.state().buttonText).toEqual('Generate');
  });

  it('should render text when no generated number data is available', () => {
    const shallowWrapper = mount(<Dashboard />);
    const renderTableSpy = spy(shallowWrapper.instance(), 'renderTable');
    shallowWrapper.setState({ data: null });

    shallowWrapper.instance().renderTable();
    expect(renderTableSpy.called).toEqual(true);
    expect(shallowWrapper.find('.empty_table__text').length).toEqual(1);
  });

  it('should call renderTable method', () => {
    const mountWrapper = mount(<Dashboard />);
    const renderTableSpy = spy(mountWrapper.instance(), 'renderTable');
    mountWrapper.setState({ data: [{ id: 1, generatedNumbers: 2340404040 }] });

    mountWrapper.instance().renderTable();
    expect(renderTableSpy.called).toEqual(true);
    expect(mountWrapper.find('Table').length).toEqual(1);
    expect(mountWrapper.find('DropDown').length).toEqual(1);
  });

  it('should render text when no generated number data is available', () => {
    const mountWrapper = mount(<Dashboard />);
    const renderTableSpy = spy(mountWrapper.instance(), 'renderTable');
    mountWrapper.setState({ data: null });

    mountWrapper.instance().renderTable();
    expect(renderTableSpy.called).toEqual(true);
    expect(mountWrapper.find('.empty_table__text').length).toEqual(1);
  });

  it('should sorts data by highest generated number', () => {
    const mountWrapper = mount(<Dashboard />);
    const handleSortClickSpy = spy(mountWrapper.instance(), 'handleSortClick');
    mountWrapper.setState({ data });
    const sortedData = _.orderBy(data, ['generatedNumbers'], ['desc']);
    mountWrapper.instance().handleSortClick(event, 'highest');
    expect(handleSortClickSpy.called).toEqual(true);
    expect(mountWrapper.state().data).toEqual(sortedData);
  });

  it('should sorts data by lowest generated number', () => {
    const mountWrapper = mount(<Dashboard />);
    const handleSortClickSpy = spy(mountWrapper.instance(), 'handleSortClick');
    mountWrapper.setState({ data });
    const sortedData = _.orderBy(data, ['generatedNumbers'], ['asc']);
    mountWrapper.instance().handleSortClick(event, 'lowest');
    expect(handleSortClickSpy.called).toEqual(true);
    expect(mountWrapper.state().data).toEqual(sortedData);
  });

  it('should sorts data by id', () => {
    const mountWrapper = mount(<Dashboard />);
    const handleSortClickSpy = spy(mountWrapper.instance(), 'handleSortClick');
    mountWrapper.setState({ data });
    const sortedData = _.orderBy(data, ['id']);
    mountWrapper.instance().handleSortClick(event, 'unique');
    expect(handleSortClickSpy.called).toEqual(true);
    expect(mountWrapper.state().data).toEqual(sortedData);
  });

  it('should validate number input field successfully', () => {
    const mountWrapper = mount(<Dashboard />);
    const validateFieldSpy = spy(mountWrapper.instance(), 'validateField');
    mountWrapper.setState({ chosenNumber: 2 });
    mountWrapper.instance().validateField();
    expect(validateFieldSpy.called).toEqual(true);
  });

  it('should validates number input field if float number is entered', () => {
    const mountWrapper = mount(<Dashboard {...props} />);
    const validateFieldSpy = spy(mountWrapper.instance(), 'validateField');
    mountWrapper.setState({ chosenNumber: 2.333 });
    mountWrapper.instance().validateField();
    expect(validateFieldSpy.called).toEqual(true);
    expect(mountWrapper.props().showMessage.called).toEqual(true);
  });

  it('should validates number input field if number greater than 500 is entered', () => {
    const mountWrapper = mount(<Dashboard {...props} />);
    const validateFieldSpy = spy(mountWrapper.instance(), 'validateField');
    mountWrapper.setState({ chosenNumber: 10000 });
    mountWrapper.instance().validateField();
    expect(validateFieldSpy.called).toEqual(true);
    expect(mountWrapper.props().showMessage.called).toEqual(true);
  });

  it('should call updateGeneratedNumbers method', () => {
    const mountWrapper = mount(<Dashboard {...props} />);
    const updateGeneratedNumbersSpy = spy(mountWrapper.instance(), 'updateGeneratedNumbers');
    mountWrapper.setState({ data });

    mountWrapper.instance().updateGeneratedNumbers();
    expect(updateGeneratedNumbersSpy.called).toEqual(true);
    expect(mountWrapper.find('Table').length).toEqual(1);
    expect(mountWrapper.find('DropDown').length).toEqual(1);
  });

  it('should call generatePhoneNumbers method', () => {
    const mountWrapper = mount(<Dashboard {...props} />);
    const generatePhoneNumbersSpy = spy(mountWrapper.instance(), 'generatePhoneNumbers');
    mountWrapper.setState({ data, choseNumber: 2 });
    expect(mountWrapper.state().data.length).toEqual(3);

    mountWrapper.instance().generatePhoneNumbers();
    expect(generatePhoneNumbersSpy.called).toEqual(true);
    expect(mountWrapper.props().showMessage.called).toEqual(true);
    expect(mountWrapper.state().data.length).toBeGreaterThanOrEqual(3);
  });

});
