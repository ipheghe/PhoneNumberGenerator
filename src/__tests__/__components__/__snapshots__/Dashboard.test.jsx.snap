// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should match component snapshot 1`] = `
<div
  className="container"
>
  <div
    className="number__input__section"
  >
    <div
      className="number__details"
    >
      <p>
        Total Numbers: 
        <span>
          0
        </span>
      </p>
      <p>
        Max Number: 
        <span>
          0
        </span>
      </p>
      <p>
        Minimum Numbers: 
        <span>
          0
        </span>
      </p>
    </div>
  </div>
  <div
    className="input__container"
  >
    <input
      className="input__field"
      id="chosenNumber"
      max="500"
      min="1"
      name="chosenNumber"
      onChange={[Function]}
      type="number"
      value={0}
    />
    <button
      className="input__button"
      onClick={[Function]}
      type="button"
    >
      Generate
    </button>
  </div>
  <div
    className="empty_table__text"
  >
    <h1>
      No Generated Number Yet.
    </h1>
    <h1>
      Please enter a number to be generated
    </h1>
  </div>
</div>
`;
